File: Week_10/Practicum/Keys and Rooms.py
def dfs(rooms, visited, current):
    if visited[current]:
        return

    visited[current] = True

    for neighb in rooms[current]:
        dfs(rooms, visited, neighb)


class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False for _ in rooms]
        
        dfs(rooms, visited, 0)

        return sum(visited) == len(rooms)

File: Week_10/Practicum/Keys and Rooms.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& rooms, int curr, vector<bool>& visited) {
        visited[curr] = true;
        for (auto& r : rooms[curr]) {
            if (!visited[r]) {
                dfs(rooms, r, visited);
            }
        }
    }
    
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        vector<bool> visited(rooms.size(), false);

        dfs(rooms, 0, visited);

        for (bool v : visited) {
            if (!v) {
                return false;
            }
        }

        return true;
    }
};